### ISSUE TITLE
Continuations run on the wrong thread

### ISSUE THREAD

**Issue by taion** [crossbario/autobahn-cpp#43]
I noticed this while thinking about #42. The current use of `boost::future::then` does not play nicely with the `asio` event loop. Specifically, `then` creates a new async thread that waits on the future, then runs the callback.

This is highly undesirable behavior. `asio` only runs its callbacks on threads that have called `io_service::run`, but using continuations on async threads in this manner entirely violate that unless the user keeps around the `io_service` and remembers to manually call `io_service::post` or something.

Even in that case, I think that the correct behavior would be to actually execute the continuation in-band immediately after resolving the promise, rather than potentially crossing thread boundaries twice to do so.

@oberstet What do you think of either explicitly taking a callback instead of returning a future, or else providing a separate e.g. `autobahn::future` class implementing `autobahn::future::then` that invokes the continuation in a correct manner? The current implementation has a pretty high risk of multithreading bugs in more complex applications.

---

I've modified the `register1` example to explicitly show this behavior by logging the thread IDs. The output looks something like:

```
Running on 105700
Starting ASIO I/O loop ..
Main thread ID 0x7fff75881300
Connected to server
Connected callback thread ID 0x7fff75881300
Session joined to realm with session ID 2018155454
Join continuation thread ID 0x101de4000
Registered with registration ID 1163389419
Registered continuation thread ID 0x101e67000
Someone is calling add2() ..
add2 invocation thread ID 0x7fff75881300
```

You can see that, while the `asio` callbacks run on the main thread as desired, the continuations do not.

**Comment by oberstet** [comment]
> Specifically, then creates a new async thread that waits on the future, then runs the callback.

This is bad. I would first try to find out if the desired behavior can be achieved using some options or whatever: have continuations run on the main thread. I can't believe no one has stepped on this before.

Rgd switching API to callbacks: no, callbacks suck and quickly lead to unmaintainable spaghetti code. I'd rather write a pure C implementation then.

**Comment by oberstet** [comment]
Can you retry what you did and specify a launch policy?

```cpp
f1.then(boost::launch::deferred, ...)
``` 

* http://stackoverflow.com/questions/22934575/why-does-boostfuturetthen-spawn-a-new-thread
* http://stackoverflow.com/questions/25327256/execute-in-main-thread-with-boostfuturethen

**Comment by taion** [comment]
`launch::deferred` isn't going to work - that just means to run the continuation in-line when you ask for the result. Since many of these continuations are used as completion handlers, they don't produce usable results, and anyway you don't want to block the IO service thread to wait anyway. In general nothing in the standard futures is going to give the right behavior - the futures need to integrate tightly with the event loop; you can't infer from fulfilling a generic promise that the right thing to do is to run the continuations immediately in-line, because that's not generally the right thing to do in a multi-threaded system.

I put up a quick example here: https://github.com/taion/AutobahnCpp/commit/2f2f8b05680ad26c67981d063afeb928b202611b. Because the result of the session join continuation is never used, that continuation never gets run, and the output is only:

```
Running on 105700
Starting ASIO I/O loop ..
Connected to server
```

-----

I think there are really 3 options here.

1. Use callbacks. This is my preferred solution. Nobody likes writing async code with callbacks, but the rest of the `asio` async function calls all use callbacks for completion handlers. Logically in many cases like for `session::join` or for making registrations, this is arguably more semantically correct anyway, since the use case you're trying to solve is to just attach a completion handler. It's the most consistent with the rest of `asio` async which most users of Autobahn|Cpp will be using anyway, so there already will be callback-based code, and it's pretty clear what's going on.

2. Use a custom future class. You've already outlined the downsides above - this is potentially very fragile.

3. Use coroutines: http://www.boost.org/doc/libs/1_58_0/doc/html/boost_asio/overview/core/spawn.html. This should give you more readable code than using callbacks, and is most likely going to work as desired, but I've never heard of anybody using coroutines in C++, and it's going to seem relatively "magical" to most developers. It does make things look more like how they'd look in Python, though.

-----

I really think the callbacks are the best option here. They're not ideal ergonomically, but it fits in with the rest of `asio` async, so at the margins you're not making things that much worse, and the other solutions are all much harder for the user to understand, which I think matters as much as making them easy to use.

**Comment by jpetso** [comment]
Preparing for C++1z, asio is making its io_service (recently renamed to io_context in asio master) into an Executor. That means it'll be possible to call `future.then(io_service, lambda);` but the downside, or bad news, is that asio in boost is not an Executor yet.

The cppwamp issue referenced above is putting a wrapper on top of io_service so it'll work as an Executor, with a bit of luck that should work for AutobahnCpp as well.

**Comment by taion** [comment]
I've long since switched to cppwamp, and given the amount of movement here since, I suspect this is no longer relevant. Closing for staleness.

**Comment by ecorm** [comment]
Asio was updated in Boost 1.66 release. There's mention of executors in this new documentation page: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/net_ts.html

I'm too busy with work/life right now to investigate any deeper. Coroutines and the traditional callback interface have been working well for us so far at work; we don't use use futures.
